Story/vspc 181 Deletion of modules

glend17 opened the pull request:
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
https://diging.atlassian.net/browse/VSPC-181
  
## Anything else the reviewer needs to know?
No

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java:
```diff
@@ -89,4 +99,32 @@ public ISequence checkIfSequenceExists(String moduleId, String sequenceId) {
+    @Override
+    public void deleteModule(String moduleId) {
+        List<ModuleLink> moduleLinksToPersist = new ArrayList<ModuleLink>();
+        if(moduleId != null) {
```
turn this logic around and return if true to get rid of this huge if block
```
if (moduleId == null) ...
```

glend17 replied:
Changed

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:
```diff
@@ -1,13 +1,21 @@
 public interface ModuleLinkRepository extends PagingAndSortingRepository<ModuleLink, String> {
+    
+    @Query("select moduleLink from ModuleLink moduleLink where module_id = ?1")
+    List<IModuleLink> findModuleLinkByModuleId(String moduleId);
```
since this method returns a list, it should use plural form `findModuleLinks...`

glend17 replied:
changed

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java:
```diff
@@ -89,4 +99,32 @@ public ISequence checkIfSequenceExists(String moduleId, String sequenceId) {
+            List<IModuleLink> moduleLinks = moduleLinkRepo.findModuleLinkByModuleId(moduleId);
+            for(IModuleLink moduleLink: moduleLinks) {
+                moduleLink.setModule(null);
+                moduleLinksToPersist.add((ModuleLink)moduleLink);
```
is there a reason you're not using the `moduleLinks` list and persist it but create a new list?

glend17 replied:
changed

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java:
```diff
@@ -89,4 +99,32 @@ public ISequence checkIfSequenceExists(String moduleId, String sequenceId) {
+            if(moduleOptional.isPresent()) {
+                Module module = moduleOptional.get();
+                List<ISlide> slides = module.getSlides();
+                for(ISlide slide: slides) {
```
merge with line 117 or even use the stream api

glend17 replied:
done

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java:
```diff
@@ -89,4 +99,32 @@ public ISequence checkIfSequenceExists(String moduleId, String sequenceId) {
+                    slideManager.deleteSlideById(slide.getId(), moduleId);
+                }
+                List<ISequence> sequences = getModuleSequences(moduleId);
+                for(ISequence sequence:sequences) {
```
same as above

glend17 replied:
done

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java:
```diff
@@ -89,4 +99,32 @@ public ISequence checkIfSequenceExists(String moduleId, String sequenceId) {
+            moduleRepo.deleteById(moduleId);
+        }
+        return;
+    }
```
you want to make sure that all of the above changes are being rolled back if an error happens along the way (we don't want half deleted modules). Either annotate this method or the whole class with (which will rollback the transaction on any exception).
```
@Transactional(rollbackFor = { Exception.class })
```

glend17 replied:
done

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteModuleController.java:
```diff
@@ -0,0 +1,37 @@
+        try {
+            moduleManager.deleteModule(moduleId);
+        }
+        catch (Exception exception) {
```
`deleteModule` doesn't seem to be throwing any exceptions. What are you handling here?

glend17 replied:
Added illegalstateexception in case module is not found

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/modules/module.html:
```diff
@@ -20,6 +20,29 @@
+	            var slides = document.getElementById("slides");
+	            startSequence.remove();
+	            sequences.remove();
+	            slides.remove();
```
does this mean the user ends up on a module page for a module that doesn't exist anymore? That doesn't make sense. They should be redirected to the module list page.

glend17 replied:
Have redirected now

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/modules/module.html:
```diff
@@ -312,7 +356,30 @@ <h4 class="modal-title" id="deleteModalTitle">Confirm Deletion?</h4>
+		    <div class="modal-dialog">
+		        <div class="modal-content">
+		            <div class="modal-header">
+		                <h4 class="modal-title" id="deleteModalTitle">Confirm Deletion?</h4>
```
no question mark at the end

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/modules/module.html:
```diff
@@ -312,7 +356,30 @@ <h4 class="modal-title" id="deleteModalTitle">Confirm Deletion?</h4>
+		            </div>
+		            <div class="modal-body">
+		                <p>
+		                    Are you sure you want to delete this module?
```
Should say:
Are you sure you want to delete this module? All content will be deleted and all module links to this module will not work anymore.

glend17 replied:
Done

jdamerow commented:
Resolve conflicts please

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleManager.java:
```diff
@@ -24,4 +25,6 @@
     
     Page<IModule> findByNameOrDescription(Pageable requestedPage,String searchText);
+
+    void deleteModule(String moduleId) throws ModuleNotFoundException;
```
Exceptions should be in the exception package inside `core`. Not sure how this one (and others) ended up in the web package, but it should be moved into the core package (it's thrown from within service package after all). Also, please move all other exceptions that are not specifically only used in the web package.

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleManager.java:
```diff
@@ -89,4 +99,34 @@ public ISequence checkIfSequenceExists(String moduleId, String sequenceId) {
+            throw new ModuleNotFoundException("Module not found");
+        }
+        moduleRepo.deleteById(moduleId);
+        return;
```
not needed

jdamerow commented:
Make it so, Jenkins.

jdamerow commented:
Same as with the other PRs., once the dependency for `@PostContruct` has been added, the build should succeed.

jdamerow commented:
Make it so, Jenkins.

glend17 commented:
Addressed pr comments, deleting all module links associated with the module as well because otherwise the loading bar in the space prevents other links from being visible

jdamerow commented:
Resolve conflicts please

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/modules/module.html:
```diff
@@ -20,6 +20,25 @@
+	        	let index = window.location.href.lastIndexOf("/");
+	        	let url = window.location.href.substring(0,index);
+	        	let redirectURL = url.replace("module","module/list");
+	        	window.location.href=redirectURL;
```
why don't you use thymeleaf to generate the url (like in line 25)?

jdamerow commented:
Make it so, Jenkins.

jdamerow commented:
after that should be good to go

glend17 commented:
Raised PR after addressing review comment

jdamerow commented:
Make it so, Jenkins.
